### PR TITLE
EKF2: ellipsoidal earth navigation

### PR DESCRIPTION
**Pull request by bresch**
### Solved Problem
The current EKF evolves in a fixed local NED tangent-plane frame with an origin set at boot. This works well for short and moderate distances but degrades when flying over long distances (>100km) as the tangent frame gets significantly further from the ellipsoid surface.

### Solution
Mechanize the navigation equations using a local-navigation-frame which is NED-oriented and with its origin attached at the CG of the vehicle. The position state now defines the global position of the drone (geodetic Latitude-Longitude-Altitude) and the local position is constructed by projecting the global position using an arbitrary origin (initialized at boot or manually by the user). 

The error-state is defined in a fixed local-navigation-frame to ignore all the -really small- pseudo-forces due to the frame rotation. This allows us to keep the covariance prediction really simple and unchanged from the previous local tangent frame approximation. The position uncertainty is also defined in the local NED which makes it more intuitive than latitude/longitude errors.

Note that the global position state is not directly included in the state vector from the derivation as lat/lon require double precision. The global position is stored in a new class `LatLonAlt` which contains operators to combine ellipsoidal position and Cartesian position errors (to compute position innovations and apply corrections). 

When no latitude/longitude/altitude is available, the EKF assumes to be started at 0/0/0. This avoids having to handle multiple conditions everywhere in the code and the incorrect radii of curvatures is insignificant when navigating close to the starting location. It is then still possible to manually set the current position of the drone to automatically initialize the EKF at the correct location and continue dead-reckoning from there.

image

The position of the output predictor is considered as an error relative to the main (delayed-time EKF) and thus integrates the velocity and position directly on the Cartesian local-navigation-frame (origin at the LLA of the delayed-time EKF).
image

### Changelog Entry
For release notes:
```
TBD
New parameter: -
Documentation: -
```

### Alternatives
-

### Test coverage
-

### Context
-

### CODE REVIEW

File: src/modules/ekf2/EKF/output_predictor/output_predictor.cpp [hunk 1]
```diff
@@ -325,7 +316,7 @@ void OutputPredictor::correctOutputStates(const uint64_t time_delayed_us,
 	// using a PD feedback tuned to a 5% overshoot
 	const float vert_vel_correction = vert_vel_integ_err * pos_gain + vert_vel_err * vel_gain * 1.1f;
 
-	applyCorrectionToVerticalOutputBuffer(vert_vel_correction);
+	applyCorrectionToVerticalOutputBuffer(vert_vel_correction); //TODO: handle ref change
```
Comment (bresch) [inline src/modules/ekf2/EKF/output_predictor/output_predictor.cpp:319]: to fix

File: src/modules/ekf2/EKF/ekf.cpp [hunk 2]
```diff
@@ -427,7 +425,7 @@ void Ekf::print_status()
 
 	printf("Position (%d-%d): [%.3f, %.3f, %.3f] var: [%.1e, %.1e, %.1e]\n",
 	       State::pos.idx, State::pos.idx + State::pos.dof - 1,
-	       (double)_state.pos(0), (double)_state.pos(1), (double)_state.pos(2),
+	       (double)_state.pos(0), (double)_state.pos(1), (double) - _gpos.altitude(),
```
Comment (bresch) [inline src/modules/ekf2/EKF/ekf.cpp:428]: xy states are always 0, display lat/lon instead or local position

File: src/modules/ekf2/EKF/ekf.h [hunk 3]
```diff
@@ -44,6 +44,7 @@
 #define EKF_EKF_H
 
 #include "estimator_interface.h"
+#include "px4_platform_common/defines.h"
```
Comment (bresch) [inline src/modules/ekf2/EKF/ekf.h:47]:
```suggestion
```

File: src/modules/ekf2/EKF/ekf_helper.cpp [hunk 4]
```diff
@@ -85,156 +85,170 @@ bool Ekf::checkAltitudeValidity(const float altitude)
...
 	}
 
+	// Reset the timout timer
+	_time_last_hor_pos_fuse = _time_delayed_us;
+
 	return true;
 }
 
-bool Ekf::setAltOriginFromCurrentPos(const float altitude, const float epv)
+bool Ekf::resetAltitudeTo(const float altitude, const float vpos_var)
```
Comment (bresch) [inline src/modules/ekf2/EKF/ekf_helper.cpp:219]: find better names than resetAltitudeTo and resetHeightTo; the difference is not directly clear

File: src/modules/ekf2/EKF/lat_lon_alt/lat_lon_alt.hpp [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+					    double &transverse_radius_of_curvature)
+	{
+		const double tmp = 1.0 - pow(Wgs84::eccentricity * sin(latitude), 2);
+		const double sqrt_tmp = std::sqrt(tmp);
+		meridian_radius_of_curvature = Wgs84::meridian_radius_of_curvature_numerator / (tmp * tmp * sqrt_tmp);
+		transverse_radius_of_curvature = Wgs84::equatorial_radius / sqrt_tmp;
+	}
+
+	struct Wgs84 {
+		static constexpr double equatorial_radius = 6378137;
```
Comment (bresch) [inline src/modules/ekf2/EKF/lat_lon_alt/lat_lon_alt.hpp:146]:
```suggestion
		static constexpr double equatorial_radius = 6378137.0;
```

**Review by Jaeyoung-Lim — COMMENTED** [review]
Nice! 

The height definitions in mavlink is based on AMSL, which are defined by a geoid. 

Wouldn't this have implications in the current navigation stack where it doesn't distinguish ellipsoidal or geoid heights?

**Comment by bresch** [comment]
> Wouldn't this have implications in the current navigation stack where it doesn't distinguish ellipsoidal or geoid heights?

@Jaeyoung-Lim You're right, it's not ideal but we still use AMSL everywhere so it's at least consistent on the altitude reporting side. A current issue is that we use AMSL instead of ellipsoid height in the calculations inside the `LatLonAlt` class.
My goal is to only use ellipsoid altitude inside EKF2 and then convert the output to AMSL. We currently don't have a geoid model, so either we add one or we continue to steal it from the GNSS receiver (i.e.: compute the difference between reported AMSL and ellipsoid alt).
This PR is however large enough so I'll do the changes in a follow-up one.

**Comment by Jaeyoung-Lim** [comment]
@bresch Thanks! I think this would be anyway a huge improvement in navigation performance, especially when you care about altitude above terrain.

Last thing I would like to comment (For myself) is that this would probably break the gazebo gps plugin setup. I have a separate plugin that considers the simulated world as the projected surface of the geoid to simulate flying in a CH1903/LV03 digital elevation map in Switzerland. We can probably adapt this to UTM coordinates for more general locations.

**Comment by bresch** [comment]
Compiled for fmu-v5
Main:
FLASH_AXIM:     2029185 B      2016 KB     98.29%

This PR:
FLASH_AXIM:     2031377 B      2016 KB     98.40% (+2192 B)

File: src/modules/ekf2/EKF/lat_lon_alt/lat_lon_alt.cpp [hunk 6]
```diff
@@ -41,10 +41,11 @@ Vector3f LatLonAlt::computeAngularRateNavFrame(const Vector3f &v_ned) const
...
 	double r_e;
 	computeRadiiOfCurvature(_latitude_rad, r_n, r_e);
+	const double alt = static_cast<double>(_altitude);
 	return Vector3f(
-		       v_ned(1) / (static_cast<float>(r_e) + _altitude),
-		       -v_ned(0) / (static_cast<float>(r_n) + _altitude),
-		       -v_ned(1) * tanf(_latitude_rad) / (static_cast<float>(r_e) + _altitude));
+		       static_cast<float>(static_cast<double>(v_ned(1)) / (r_e + alt)),
+		       static_cast<float>(static_cast<double>(-v_ned(0)) / (r_n + alt)),
+		       static_cast<float>(static_cast<double>(-v_ned(1)) * std::tan(_latitude_rad) / (r_e + alt)));
```
Comment (bresch) [inline src/modules/ekf2/EKF/lat_lon_alt/lat_lon_alt.cpp:48]: `std::tan(double)` is too expensive -> remove

**Comment by hamishwillee** [comment]
@bresch Would appreciate your thoughts on this one: https://github.com/mavlink/mavlink/pull/2168#issuecomment-2489589917 - we're proposing that MAVLink uses  EGM96 geoid by default as the global altitude reference instead of "UBLOX MSL". Wondering if you might be able to common on the implications.

File: src/modules/ekf2/EKF/ekf.h [hunk 7]
```diff
@@ -711,14 +711,22 @@ class Ekf final : public EstimatorInterface
...
+
+	Vector2f computeDeltaHorizontalPosition(const double &new_latitude, const double &new_longitude) const;
+	void updateHorizontalPositionResetStatus(const Vector2f &delta);
 
 	void resetWindTo(const Vector2f &wind, const Vector2f &wind_var);
 
 	bool isHeightResetRequired() const;
 
-	void resetVerticalPositionTo(float new_vert_pos, float new_vert_pos_var = NAN);
+	void resetHeightTo(float new_altitude, float new_vert_pos_var = NAN);
```
Comment (haumarco) [inline src/modules/ekf2/EKF/ekf.h:728]: Can we check again the `resetHeightTo` naming and usage

**Review by haumarco — APPROVED** [review]
Looking good, since all performance tests in SITL and SIH were successful it should be ready.

**Pull request merged**